pkg/shared/utils/arrays/generic_array_utils.go

```diff
@@ -1,12 +1,86 @@
 package arrays
 
+const NotFound = -1
+
+type PredicateFunc[K any] func(x K) bool
+type ReducerFunc[K any, V any] func(accumulatorValue V, currentValue K) V
+type MapperFunc[K any, V any] func(x K) V
+type ConsumerFunc[K any] func(x K)
+
 // FindFirst Finds the first element into the passed in values where the predicate is true
-// WARNING: passed in values are converted to an []interface{} slice and returned value will be an interface{}
-func FindFirst(predicate func(x interface{}) bool, values ...interface{}) (int, interface{}) {
+func FindFirst[K any](values []K, predicate PredicateFunc[K]) (int, K) {
 	for i, val := range values {
 		if predicate(val) {
 			return i, val
 		}
 	}
-	return -1, nil
+	var res K
+	return NotFound, res
+}
+
+// Filter return a slice containing only the values of `values` matching the given predicate
+func Filter[K any](values []K, predicate PredicateFunc[K]) []K {
+	var res []K
+	for _, val := range values {
+		val := val
+		if predicate(val) {
+			res = append(res, val)
+		}
+	}
+	return res
+}
+
+// Map - creates a new slice where each element is the result of the mapper function to the corresponding element in the input slice
+func Map[K any, V any](values []K, mapper MapperFunc[K, V]) []V {
+	var res []V
+	for _, val := range values {
+		res = append(res, mapper(val))
+	}
+	return res
+}
+
+// AnyMatch - check whether at least one of the element of the given slice matches the given predicate
+func AnyMatch[K any](values []K, predicate PredicateFunc[K]) bool {
+	if idx, _ := FindFirst(values, predicate); idx != -1 {
+		return true
+	}
+	return false
+}
+
+// NoneMatch - check that none of the elements of the given slice matches the given predicate
+func NoneMatch[K any](values []K, predicate PredicateFunc[K]) bool {
+	return !AnyMatch(values, predicate)
+}
+
+// AllMatch - check if all the elements of the given slice matches the given predicate
+func AllMatch[K any](values []K, predicate PredicateFunc[K]) bool {
+	for _, val := range values {
+		if !predicate(val) {
+			return false
+		}
+	}
+	return true
+}
+
+// Reduce - reduces an array to a single value by applying the reducer function.
+// initialValue is the starting value of the reduction
+func Reduce[K any, V any](values []K, reducer ReducerFunc[K, V], initialValue V) V {
+	res := initialValue
+	for _, val := range values {
+		val := val
+		res = reducer(res, val)
+	}
+	return res
+}
+
+// Contains - check whether the given array contains the given value
+func Contains[K comparable](values []K, s K) bool {
```

Comment:
I think we should probably expand documentation around this method about how it specifically checks that, and also remark the difference with AnyMatch.
Specifically, this compares using the `==` operator. People might be wondering when to use `Contains` and when to use `AnyMatch` thinking it performs some kind of semantic comparison whereas it just uses the `==` operator.